libp2p-comit/src/substream/inbound.rs

```diff
@@ -5,50 +5,52 @@ use crate::{
     substream::{Advance, Advanced, CloseStream},
     Frame, FrameType, IntoFrame,
 };
-use futures::sync::oneshot;
-use libp2p_swarm::ProtocolsHandlerEvent;
-use std::collections::{HashMap, HashSet};
-use tokio::prelude::*;
+use futures::{channel::oneshot, task::Poll, Future, Sink, Stream};
+use libp2p::swarm::ProtocolsHandlerEvent;
+use std::{
+    collections::{HashMap, HashSet},
+    pin::Pin,
+    task::Context,
+};
 
 #[derive(strum_macros::Display)]
 #[allow(missing_debug_implementations)]
 /// States of an inbound substream i.e. from peer node to us.
-pub enum State<TSubstream> {
+pub enum State {
     /// Waiting for a request from the remote.
-    WaitingMessage { stream: Frames<TSubstream> },
+    WaitingMessage { stream: Pin<Box<Frames>> },
```

Comment:
We have to `Pin` the stream to a specific location in memory.

The current code here does what async/await would do under the hood: It represents the state as an enum and each variant is a check (await) point.

async functions automatically use `Pin` if necessary.
We have to use it here, to be able to access the `poll` APIs on the stream.